micro-rdk/src/common/provisioning/server.rs

```diff
@@ -0,0 +1,528 @@
+        debug_assert_eq!(buffer.len(), 5 + len);
+        debug_assert_eq!(buffer.capacity(), 5 + len);
+        Ok(buffer.freeze())
+    }
+
+    fn set_smart_machine_credentials(&self, body: Bytes) -> Result<Bytes, GrpcError> {
+        let creds =
+            SetSmartMachineCredentialsRequest::decode(body).map_err(|_| GrpcError::RpcInternal)?;
+        self.storage.store_robot_credentials(creds.cloud.unwrap())?;
+        let resp = SetSmartMachineCredentialsResponse::default();
+
+        let len = resp.encoded_len();
+        let mut buffer = BytesMut::with_capacity(5 + len);
+        buffer.put_u8(0);
+        buffer.put_u32(len.try_into().unwrap());
+        resp.encode(&mut buffer)
+            .map_err(|_| GrpcError::RpcInternal)?;
+        debug_assert_eq!(buffer.len(), 5 + len);
+        debug_assert_eq!(buffer.capacity(), 5 + len);
+
+        Ok(buffer.freeze())
+    }
+
+    async fn process_request(
+        &self,
+        req: Request<Incoming>,
+    ) -> Result<Response<GrpcBody>, http::Error> {
+        let mut resp = GrpcBody::new();
+        match self.process_request_inner(req).await {
+            Ok(bytes) => resp.put_data(bytes),
+            Err(e) => resp.set_status(e.to_status("".to_string()).code, None),
+        };
+
+        Response::builder()
+            .status(200)
+            .header(CONTENT_TYPE, "application/grpc")
+            .body(resp)
+    }
+
+    fn get_credential_ready(&self) -> AtomicSync {
+        self.credential_ready.clone()
+    }
+}
+
+impl<S> Service<Request<Incoming>> for ProvisioningService<S>
+where
+    S: Storage + Clone + 'static,
+    GrpcError: From<S::Error>,
+{
+    type Response = Response<GrpcBody>;
+    type Error = http::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
+    fn call(&self, req: Request<Incoming>) -> Self::Future {
+        let svc = self.clone();
+        Box::pin(async move { svc.process_request(req).await })
+    }
+}
+#[pin_project::pin_project]
+struct ProvisoningServer<I, S, E>
+where
+    S: Storage + Clone + 'static,
+    GrpcError: From<S::Error>,
+{
+    _exec: PhantomData<E>,
+    _stream: PhantomData<I>,
+    _storage: PhantomData<S>,
+    #[pin]
+    connection: http2::Connection<I, ProvisioningService<S>, E>,
+    credential_ready: AtomicSync,
+}
+
+impl<I, S, E> Future for ProvisoningServer<I, S, E>
+where
+    S: Storage + Clone + 'static,
+    I: rt::Read + rt::Write + std::marker::Unpin + 'static,
+    GrpcError: From<S::Error>,
+    E: rt::bounds::Http2ServerConnExec<
+        <ProvisioningService<S> as Service<Request<Incoming>>>::Future,
+        GrpcBody,
+    >,
+{
+    type Output = Result<(), hyper::Error>;
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let this = self.project();
+
+        match this.connection.poll(cx) {
+            Poll::Pending => futures_lite::ready!(this.credential_ready.poll(cx)),
```

Comment:
At this time it's unclear whether we should shutdown provisioning when credential have been stored or if we let the provisioning client close the connection.
Since we do not envision running the provisioning server while the main server is running we will have to pick a poison :)

Reply:
I shudder to think that this is possible, but could this not potentially be used in the future to reconfigure the robot (in which case we would not want to shut it down)?